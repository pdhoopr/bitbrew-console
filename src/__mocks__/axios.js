function getDefaults({ baseURL = '', headers = {} } = {}) {
  return {
    baseURL,
    headers: {
      common: headers,
    },
  };
}

function mockHttpMethod() {
  return jest.fn().mockResolvedValue({
    data: {},
  });
}

class AxiosMock {
  defaults = getDefaults();
  get = mockHttpMethod();
  post = mockHttpMethod();
  put = mockHttpMethod();
  delete = mockHttpMethod();
}

class ApiMock {
  api = new AxiosMock();

  create = jest.fn((config) => {
    this.api.defaults = getDefaults(config);
    return this.api;
  });

  reset = () => {
    this.api = new AxiosMock();
  };
}

export default new ApiMock();
