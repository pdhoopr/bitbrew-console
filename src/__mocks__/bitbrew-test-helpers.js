import { shallow } from 'enzyme';
import enzymeSerializer from 'enzyme-to-json/serializer';
import diff from 'jest-diff';
import { addSerializer, getSerializers, utils } from 'jest-snapshot';
import styledSerializer from 'jest-styled-components/src/styleSheetSerializer';
import stripAnsi from 'strip-ansi';

const DIFF_HEADER = 'Diff:\n';

export const diffComponents = (baseComponent, compareComponent) => {
  const serializers = getSerializers();
  [styledSerializer, enzymeSerializer].forEach(serializer => {
    if (!serializers.includes(serializer)) {
      addSerializer(serializer);
    }
  });
  const base = `${utils.serialize(shallow(baseComponent))}\n`;
  const compare = `${utils.serialize(shallow(compareComponent))}\n`;
  const changes = diff(base, compare, {
    aAnnotation: `<BaseComponent />`,
    bAnnotation: `<CompareComponent />`,
    contextLines: 3,
    expand: false,
  });
  return `${DIFF_HEADER}${changes}`;
};

export const diffSerializer = {
  test(value) {
    return typeof value === 'string' && value.startsWith(DIFF_HEADER);
  },
  print(value) {
    return stripAnsi(value).trim();
  },
};

export const fixtures = {
  accessToken:
    'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE1MjUxMTUzNTAsImV4cCI6MTU1NjY1MTM1MCwiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSJ9.4O-W8aeOOUFU8fkGTAFaAu9D9nK_l61yPomNG1Qkny8',
  orgs: {
    danlaw: {
      id: 'bd006616-3bd2-4783-a84d-b0fa4ed81135',
      properName: 'Danlaw',
    },
    azuga: {
      id: '8d9ec637-f3bb-4c46-83ee-11764d09f0b2',
      properName: 'Azuga',
    },
    bitbrew: {
      id: '94d9b0ca-a758-4502-8dcd-41d883e38714',
      properName: 'BitBrew',
    },
  },
  users: {
    bill: {
      email: 'bill@bitbrew.com',
      name: 'Bill Monaghan',
    },
    cynthia: {
      email: 'cynthia@bitbrew.com',
      name: 'Cynthia Hamilton',
    },
    alex: {
      email: 'bill@bitbrew.com',
      name: 'Alex Li',
    },
  },
};
