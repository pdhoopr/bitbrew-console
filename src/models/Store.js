import { autorun } from 'mobx';
import { flow, types } from 'mobx-state-tree';
import api from '../utils/api';
import Org from './Org';

export default types
  .model('Store', {
    orgs: types.optional(types.array(Org), []),
    storageKey: types.optional(types.string, 'bitbrew-console'),
    token: types.maybe(types.string),
  })
  .views(self => ({
    get isSignedIn() {
      return !!self.token;
    },
    get alphabetizedOrgs() {
      return [...self.orgs].sort((a, b) =>
        a.properName.toLowerCase().localeCompare(b.properName.toLowerCase()),
      );
    },
  }))
  .actions(self => ({
    afterCreate() {
      self.token = localStorage.getItem(self.storageKey);
    },
    signIn({ accessToken }) {
      self.token = accessToken;
      localStorage.setItem(self.storageKey, accessToken);
    },
    signOut() {
      self.token = null;
      localStorage.removeItem(self.storageKey);
    },
    configureApi: autorun(() => {
      api.configure({
        token: self.token,
      });
    }),
    createOrg: flow(function* createOrg(data) {
      yield api.orgs.create(data);
    }),
    listOrgs: flow(function* listOrgs() {
      const { items } = yield api.orgs.list();
      self.orgs = items;
    }),
  }));
