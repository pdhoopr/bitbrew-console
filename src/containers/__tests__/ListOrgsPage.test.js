import axios from 'axios';
import { fixtures } from 'bitbrew-test-helpers';
import { shallow } from 'enzyme';
import { Provider } from 'mobx-react';
import React from 'react';
import { render, wait } from 'react-testing-library';
import Store from '../../models/Store';
import ConnectedListOrgsPage from '../ListOrgsPage';

const ListOrgsPage = ConnectedListOrgsPage.wrappedComponent;

beforeEach(() => {
  localStorage.clear();
  axios.reset();
});

test('renders DOM for the list orgs page', () => {
  const dom = shallow(
    <ListOrgsPage listOrgs={jest.fn()} orgs={[]} signOut={jest.fn()} />,
  );
  expect(dom).toMatchSnapshot();
});

test('can successfully list orgs from the API', async () => {
  const items = Object.values(fixtures.orgs);
  axios.api.get.mockResolvedValueOnce({
    data: { items },
  });
  const { getAllByTestId, getByTestId, getByText } = render(
    <Provider store={Store.create()}>
      <ConnectedListOrgsPage />
    </Provider>,
  );
  expect(getByText('Welcome!')).toBeDefined();
  expect(() => getByTestId('org-name')).toThrow();
  await wait();
  const allOrgNames = getAllByTestId('org-name').map(
    orgName => orgName.textContent,
  );
  expect(allOrgNames.length).toBe(items.length);
  items.forEach(item => {
    expect(allOrgNames).toContain(item.properName);
  });
});
