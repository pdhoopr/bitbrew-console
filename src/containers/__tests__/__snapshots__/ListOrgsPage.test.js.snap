// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DOM for the list orgs page 1`] = `
<React.Fragment>
  <styled.header>
    <styled.div>
      <styled.div>
        <styled.svg />
        <styled.button
          onClick={[MockFunction]}
        >
          Sign out
        </styled.button>
      </styled.div>
      <styled.h1>
        Welcome!
      </styled.h1>
      <styled.p>
        Use the Dashboard to manage your data pipeline.
      </styled.p>
    </styled.div>
  </styled.header>
  <styled.main>
    <styled.div>
      <OrgList
        orgs={Array []}
      />
    </styled.div>
  </styled.main>
  <Footer />
</React.Fragment>
`;
